dask_expr/_expr.py

```diff
@@ -2190,18 +2215,24 @@ def optimize(expr: Expr, combine_similar: bool = True, fuse: bool = True) -> Exp
     optimize_blockwise_fusion
     """
 
-    result = expr
-    while True:
-        out = result.simplify().lower_once()
-        if out._name == result._name:
-            break
-        result = out
+    # Simplify
+    result = expr.simplify()
 
+    # Combine similar
     if combine_similar:
         result = result.combine_similar()
 
+    # Manipulate Expression to make it more efficient
+    result = result.rewrite(kind="tune")
```

Comment:
Nice.

I find it interesting that the "tune" step comes before `lower_completely`, given that this kind of optimization is related to the partition count. My understanding is that we **need** to do this after `combine_similar`. Do we also **need** to do this before `lower_completely`? Or, is the idea that you might as well aggregate partitions before the expression graph gets larger?

Note that I don't mind the ordering you have here, but I'm slightly worried if it **needs** to be that way for some reason.

Reply:
It definitely needs to be in that order. Tune (open for other names btw) changes the number of partitions while lower locks the number of partitions in. We run into all kinds of trouble if we lock them in and then try to change them afterwards

Reply:
Ah okay, that makes sense! Sorry for the confusion.

Personal Notes: This optimization lives in that "fuzzy zone" where you don't really care if you have "lowered" expressions, but you cannot have "frozen" partitioning anywhere.  Right now, a "lowered" expression is just one that can directly generate a task graph, which is slightly different from meaning that it has frozen partitions. However, we have certainly avoided frozen partitioning outside of lowered expressions so far (so this ordering makes sense).